### ISSUE TITLE
Aliases contain many invalid entries

### ISSUE THREAD

**Issue by cristan** [datasets/un-locode#27]
As far as I can tell: these aliases are useless:

```
GL,Fredrikshaab = Paamiut,Fredrikshaab = Paamiut
GL,Godhavn = Qeqertarsuaq,Godhavn = Qeqertarsuaq
GL,Gronnedal = Kangilinguit,Gronnedal = Kangilinguit
```

etc. See this logs:
```
Sisimiut in GL not found!
Ilulissat in GL not found!
Qaqortoq in GL not found!
Illorqortoormiut in GL not found!
Kangerlussua in GL not found!
Manitsoq in GL not found!
Bishkek in KG not found!
Thandwe in MM not found!
Bagan Luar in MY not found!
Nizhny Novgorod in RU not found!
La Brea in TT not found!
Adak Island in US not found!
Sibbo (Sipoo) in FI not found!
Sibbo (Sipoo) in FI not found!
Vanda (Vantaa) in FI not found!
Chuuk in FM not found!
Fuglafirdi in FO not found!
Paamiut in GL not found!
Qeqertarsuaq in GL not found!
Kangilinguit in GL not found!
Sisimiut in GL not found!
Ilulissat in GL not found!
Qaqortoq in GL not found!
Illorqortoormiut in GL not found!
Kangerlussua in GL not found!
Manitsoq in GL not found!
Bishkek in KG not found!
Thandwe in MM not found!
Bagan Luar in MY not found!
Nizhny Novgorod in RU not found!
La Brea in TT not found!
Adak Island in US not found!
Sibbo (Sipoo) in FI not found!
Sibbo (Sipoo) in FI not found!
Vanda (Vantaa) in FI not found!
Chuuk in FM not found!
Fuglafirdi in FO not found!
```

This is a working alias:
```
BE,Antwerp = Antwerpen,Antwerp = Antwerpen
```
This is useful: when searching for Antwerp, you actually want Antwerpen ([BEANR](https://unlocode.info/BEANR)).

Compare this with this one:
```
GL,Fredrikshaab = Paamiut,Fredrikshaab = Paamiut
```
This is quite useless: the name of [GLJFR](https://unlocode.info/GLJFR) is Paamiut (Fredrikshaab), so I can't tell which UN/LOCODE is meant. Or am I missing something? Many of the incorrect ones are added in the latest release.

**Comment by sabas** [comment]
alias.csv is created by stripping the rows with = as status in the original dataset, perhaps the rows were kept while the name changed in one of the latest releases?

**Comment by cristan** [comment]
No, that's not it. Many aliases [are changed in 2024-1](https://github.com/datasets/un-locode/commit/600bdf54c4777a7328123d16166b1f8bbc7ee359), so they are as new as they can be and many of these are wrong. In fact, it already goes wrong at the second new entry:

```
GL,Fredrikshaab = Paamiut,Fredrikshaab = Paamiut
```

However GLJFR has the name _Paamiut (Fredrikshaab)_ and that's not in the aliases.

**Comment by eromoe** [comment]
@cristan May I ask which un-locode dataset you are using now ?

**Comment by eromoe** [comment]
@cristan Hi, I see your fork https://github.com/cristan/improved-un-locodes , does it solve this aliases issue?

**Comment by cristan** [comment]
@eromoe I'm afraid I didn't fix the aliases in my fork, no. I could do it though if you want to: it probably won't be that large amount of work. The question is: what good would it do? 

As of 2024-2, there are just about 116.000 UN/LOCODEs and only 86 aliases. That is such a limited dataset that you can question whether you should even bother with such a small list. It most definitely isn't a comprehensive of alternate spellings. For example: [BEBGS](https://unlocode.info/BEBGS) has as its name Brugge (Bruges) (so basically has 2 spellings of its name) and [ITVCE](https://unlocode.info/ITVCE) has the name Venezia while there isn't an alias also naming it Venice.

If you want aliases, you'd probably be better off using another dataset. You can find one or could be a fun project to start by yourself. A source could be Wikidata. If you check out https://www.wikidata.org/wiki/Q641, you'll see that it has different spellings of Venice already specified, as well as it having the un/locode tag with the value ITVCE, so that should be all you need to get going.

**Comment by eromoe** [comment]
I see, thank you very much for detailed explanation.

**Comment by cristan** [comment]
@eromoe Well are you in luck: it just happened to be that I needed the aliases myself for my job, so I have created (I think) exactly what you were looking for: [aliases-improved.csv](https://github.com/cristan/improved-un-locodes/blob/main/data/aliases-improved.csv).

I think it's tremendously better than what UN/LOCODE has:
* Much more user friendly IMO. Example: `ITMIL,Milan` as opposed to `IT,Milan = Milano,Milan = Milano`. So now you don't have to parse the string and look up the name to find out what UN/LOCODE it's an alias for.
* All the mappings actually work
* Better mappings: `Amoy = Xiamen Gaoqi International Apt` makes no sense: Amoy is an alias for Xiamen, the airport has nothing to do with it. `AT,Vienna = Wein,Vienna = Wein` is plain incorrect, it should be `Wien`.
* And last but not least: way more of them. The official dataset has 86 aliases. Mine has more than 575.000.

Let me know if this works for you.

Shameless plug: I built this as a new feature for [our UN/LOCODE API](https://www.portbase.com/marketplace/cargoprobe-un-locode-api/). We already had the improved UN/LOCODEs as a convenient API, but now you can also find the correct UN/LOCODE + its improved data with a city + (optional) country.

**Comment by eromoe** [comment]
Cool ! Include many translation, it is very helpful for internationalization after human review .
Now I am in a very different situation that struggling mess of local `native LOCODE`  which can't match with UN/LOCODE.And we usually use one LOCODE in a city though there are 6 UN/LOCODE . For example, we all use `CNSHA` for `Shanghai`  instead of `Shanghai Hongqiao International Apt`  or `Hongqiao`,  some of my customor would use `CNPVG` as `Shanghai` .
![Image](https://github.com/user-attachments/assets/6e776d05-42aa-48b5-b910-ff1b3f2be7d4)

The native usage is a big problem.

**Comment by sabas** [comment]
@eromoe the CNSHA issue is discussed in the community since long years hahaha
In the maritime we keep using CNSHA

**Comment by cristan** [comment]
Yeah, that's a problem, and one which many people have. There were talks of a project where a hierarchies would be defined. I think it would look something like this:

```
Unlocode,Parent
CNSHA,CNSGH
CNPVG,CNPDG
CNSHZ,CNSGH
CNPDG,CNSGH
```

This way you would know these are all in Shanghai. I think the project got cancelled though because funding got cut.

Personally, I'd move the UN/LOCODE dataset to GitHub, create a validation script that any entries in this list point to real UN/LOCODES and let the community solve it. There are talks of going the open source route, but I wouldn't hold my breath.

Anyway, luckily you have me 😁 Some vibe coding later and I now have [parents.csv](https://github.com/cristan/improved-un-locodes/blob/main/data/parents.csv). This might be what you need.

**Comment by eromoe** [comment]
Looks nice! Do you extract the location(city/provice) from Name and use the location to build the mapping ?

I think the most important mapping is location, because many people only use location to identify locode, I also wrote some rule to parse the location
![Image](https://github.com/user-attachments/assets/832c29d4-0e5d-4cdd-8a92-b7cdb3fc76ae)
, and may use ai double check in future.

[aliases-improved.csv](https://github.com/cristan/improved-un-locodes/blob/main/data/aliases-improved.csv) is very useful for ports . In my case, we need alias for location over ports, do you have a recommended dataset ?

**Comment by cristan** [comment]
> Looks nice! Do you extract the location(city/provice) from Name and use the location to build the mapping ?

No, I did not. That only gets you so far. For example this line:

```
DEBHQ,DEHAM
```
[DEBHQ](https://unlocode.info/DEBHQ) is Bahrenfeld, but you wouldn’t know from the name alone that it's part of [DEHAM](https://unlocode.info/DEHAM), which is Hamburg.

Instead, I used Wikidata. Here, [Bahrenfeld](https://www.wikidata.org/wiki/Q1608) is located in the administrative territorial entity Altona, which is located in the administrative territorial entity [Hamburg](https://www.wikidata.org/wiki/Q1055).

I've been thinking, dependent on what you need, you can even skip using parents.csv completely. In [code-list-improved.csv](https://raw.githubusercontent.com/cristan/improved-un-locodes/refs/heads/main/data/code-list-improved.csv), you'll see DEBHQ is at `53.55000,9.91667` and DEHAM is at `53.51667,9.93333`. That’s only 3.9 km apart, so depending on your use case, you could treat that as “close enough.”

> [aliases-improved.csv](https://github.com/cristan/improved-un-locodes/blob/main/data/aliases-improved.csv) is very useful for ports . In my case, we need alias for location over ports, do you have a recommended dataset ?
I don't know what you mean. What would that dataset look like?

**Comment by eromoe** [comment]
WOW, you give me a fantasic hint ! 
1. get administrative territorial entity list 
2. calculate the distance and get best location guess. 

Though some ports are probably closer to neighboring administrative regions than to their own need double check, really a good way by geographic coordinates, can use to recommandation too!
